### ISSUE TITLE
[Question] Extremely long build time (>50 hours) when compiling Transformer/SNN models with TensorRT

### ISSUE THREAD

**Issue by pengjh0111** [NVIDIA/TensorRT#4583]
I am a beginner with TensorRT, and I am currently trying to optimize some classic SNN (Spiking Neural Network) models using TensorRT.

When I previously used TensorRT for CNN-type networks, the scheduling/compilation process was quite fast — I could generate an .engine file within minutes.

However, when I try to import networks with transformer structures, such as BERT, the scheduling time becomes extremely long. For example, in one model I tested, the process has already taken more than 50 hours and is still not finished. Is this expected behavior?

I also noticed a similar issue mentioned in a 2020 paper, where the authors reported that scheduling LSTM networks with TensorRT also took an abnormally long time.

Since SNNs also have temporal characteristics similar to LSTMs, I wonder:
	•	Is this a known limitation or issue in TensorRT when dealing with such models?
	•	Or could it be that my scheduling code has some problem that causes this excessive compilation time?

**Comment by SumitdevelopAI** [comment]
Enable Builder Flags to Limit Search

Use builder.setFlag(trt.BuilderFlag.FP16) or INT8 (if calibrated). This reduces the number of tactics TensorRT tries.

Use builder.max_workspace_size wisely (don’t make it unnecessarily huge).

Use Heuristic Tactic Sources
TensorRT ≥ 8.0 lets you control tactic sources. For example:

config.set_tactic_sources([trt.TacticSource.CUBLAS, trt.TacticSource.CUBLAS_LT])


This avoids TensorRT exploring all kernels (which is what causes 50h builds).

Limit Timing Iterations

config.set_timing_cache(timing_cache, ignore_mismatch=True)
config.max_timing_iterations = 2
config.min_timing_iterations = 1


By default TensorRT may do hundreds of timing iterations per layer, which explodes build time.

Use Prebuilt / Pre-Optimized Plugins
For BERT/Transformer-like models, NVIDIA provides TensorRT OSS and FasterTransformer. These include hand-optimized fused kernels that skip the massive scheduling step.

If you use their plugins, build times drop from hours to minutes.

Profile the Graph
You can dump the TensorRT network to see which subgraph is consuming build time. Often it’s MatMul + Softmax in transformers, or RNNCell in LSTMs/SNNs.

**Comment by pengjh0111** [comment]
> Limit Timing Iterations  限制时间迭代次数
> 
> config.set_timing_cache(timing_cache, ignore_mismatch=True) config.max_timing_iterations = 2 config.min_timing_iterations = 1
> 


Does version 10.13.2.6 no longer support configuring the number of iterations? I got an error when I was actually using it, saying that there is no such attribute.

In addition, I have already set "tactic_sources = 1 << int(trt.TacticSource.CUBLAS)". If I change it to config.set_tactic_sources(0), does this mean disabling all libraries, which will further reduce the search space? 

Thanks for your help！
@SumitdevelopAI

**Comment by SumitdevelopAI** [comment]
In TensorRT 10.x (including 10.13.2.6) the old IBuilderConfig timing attributes such as max_timing_iterations / min_timing_iterations were removed or deprecated from the Python bindings, so assigning to them raises an AttributeError; instead use the newer timing controls (for example config.avg_timing_iterations) and — more importantly — reuse a timing cache with config.set_timing_cache(timing_cache, ignore_mismatch=True) to avoid repeated benchmarking. If you need to reduce the search space to speed up builds, prefer safe controls such as lowering avg_timing_iterations, setting conservative workspace limits via config.set_memory_pool_limit(trt.MemoryPoolType.WORKSPACE, value) (which prevents workspace-hungry tactics from being considered), or applying an IAlgorithmSelector to exclude specific algorithms. Calling config.set_tactic_sources(0) will disable all tactic providers (bitmask = 0), which does dramatically reduce search work but can also make many layers unbuildable or break plugins that expect cuBLAS/cuDNN handles — so only disable tactic sources selectively (enable only the providers you know your model supports) rather than using 0 unless you accept possible build failures.
